src/AWS.Logger.Core/Core/AWSLoggerCore.cs

```diff
@@ -175,60 +164,90 @@ public void StartMonitor()
             try
             {
                 token.ThrowIfCancellationRequested();
+
                 if (_currentStreamName == null)
                 {
-                    await LogEventTransmissionSetup(token).ConfigureAwait(false);
+                    _currentStreamName = await CreateLogStream(token).ConfigureAwait(false);
                 }
+
                 while (!token.IsCancellationRequested)
                 {
                     if (!_pendingMessageQueue.IsEmpty)
                     {
-                        while (!_pendingMessageQueue.IsEmpty)
+                        var pendingMessages = Interlocked.Exchange(ref _pendingMessageQueue, new ConcurrentQueue<InputLogEvent>());
+
+                        // If the max messages are 10k and there is 25k in the queue, it will create 10k + 10k + 5k chunks
+                        var chunks = pendingMessages
```

Comment:
Similar to comments about the extension method. This is recreating another potential large list. It would be great if we could just iterate over the pendingMessages once.

Reply:
This is something that you would do?

```csharp
while(!pendingMessages.IsEmpty)
{
    var batchList = new List<InputLogEvent>();

    while (pendingMessages.TryDequeue(out var logEvent) && batchList.Count <= _config.MaxQueuedMessages)
    {
        batchList.Add(logEvent);
    }

    var logBatch = LogEventBatch.CreateBatch(batchList, SequenceToken, _currentStreamName, _config);

    await SendMessages(logBatch, token).ConfigureAwait(false);
}
```
Or an extension method like DequeueWhile?
```csharp
var batch = pendingMessages.DequeueWhile(() => batchList.Count <= _config.MaxQueuedMessages)
```
```csharp
internal static IEnumerable<T> DequeueWhile<T>(this ConcurrentQueue<T> queue, Func<bool> predicate)
{
   while(!queue.IsEmpty && predicate())
   {
         if (queue.TryDequeue(out var item))
                 yield return item;
   }
}
```
Do you think its faster and more efficient with resources?

Reply:
Yeah I think that is close to what I mean but where does the IsSizeConstraintViolated method come into play? The `_config.MaxQueuedMessages` constraint could be larger then what PutLogEvents can take which is a total of 1 meg of data.